src/util/qstringformat.h

```diff
@@ -0,0 +1,42 @@
+#pragma once
+
+#include <QString>
+#include <type_traits>
+#include <utility>
+
+namespace {
+
+// taken from Qt
+template<typename T>
+static constexpr bool is_convertible_to_view_or_qstring_v =
+        std::is_convertible_v<T, QString> ||
+        std::is_convertible_v<T, QStringView> ||
+        std::is_convertible_v<T, QLatin1String>;
+
+// check if we can call QString::number(T) with T
+template<typename T>
+static constexpr bool is_number_compatible_v =
+        std::is_invocable_v<decltype(QString::number(std::declval<T>()))(T), T>;
+
+// always false, used for static_assert and workaround for compilers without
+// https://cplusplus.github.io/CWG/issues/2518.html
+template<typename T>
+static constexpr bool always_false_v = false;
+} // namespace
+
+// Try to convert T to a type that would be accepted by QString::args(Args&&...)
+template<typename T>
+auto convertToQStringConvertible(T&& arg) {
+    if constexpr (is_convertible_to_view_or_qstring_v<T>) {
+        // no need to do anything, just return verbatim
+        return std::forward<T>(arg);
+    } else if constexpr (is_number_compatible_v<T>) {
+        return QString::number(std::forward<T>(arg));
+    } else {
+        static_assert(always_false_v<T>, "Unsupported type for QString::arg");
```

Comment:
unfortunately this throws an error when picked for main (#12548), dunno if that is to be expected, let alone how to fix it (except using `static_assert(false, ...)` )
https://github.com/mixxxdj/mixxx/actions/runs/9103424016/job/25025246943?pr=12548
```
/src/util/stringformat.h: In function ‘auto convertToQStringConvertible(T&&)’:
Error: /src/util/stringformat.h:36:23: error: reference to ‘always_false_v’ is ambiguous
   36 |         static_assert(always_false_v<T>, "Unsupported type for QString::arg");
      |                       ^~~~~~~~~~~~~~
/src/util/stringformat.h:24:23: note: candidates are: ‘template<class T> constexpr const bool {anonymous}::always_false_v<T>’
   24 | static constexpr bool always_false_v = false;
      |                       ^~~~~~~~~~~~~~
```

Reply:
I think its getting confused because `always_false_v` is also already defined in other headers (likely from #12781). So you should be able to fix it by shuffling some code around. Potentially just putting a single definition in a single header and then using that across both files.

Reply:
Ah, okay, in `main` this seems to clash with `always_false_v` declared in midimessage.h (where it is `inline constexpr` bool)
https://github.com/mixxxdj/mixxx/blob/8f647908af460e53fe8a860b6ce0964a93a55112/src/controllers/midi/midimessage.h#L183

Reply:
(comment race condition : )

Reply:
> Potentially just putting a single definition in a single header and then using that across both files could be enough

What place would you pick?